### ISSUE TITLE
NoCFI meaning when combining policies

### ISSUE THREAD

**Issue by amstrnad** [draperlaboratory/hope-policies#85]
The `cfi` and `threeClass` policies both use `NoCFI`, but it tags slightly different instructions. When the policies are both used, the intersection of instructions that each would tag `NoCFI` are tagged.

If this is correct behavior, we might want to add a warning. If not, we should figure out what the correct behavior is and update the tagging.

To reproduce, compare the tagged binaries on [`pr-strnad-ignore`](https://github.com/draperlaboratory/hope-src/tree/pr-strnad-ignore), which has some changes to make this more visible.

**Comment by ccasin** [comment]
Oh boy - sounds like a bug.  The "right" behavior is for their to be two "NoCFI" tags.  That is, `osv.cfi.NoCFI` and `osv.threeClass.NoCFI` are separate tags represented by separate bits in the metadata sets.  They shouldn't interact in any way.

I'm confused about how these two policies are seeing this tag applied to different instructions.  It looks like they both have the same initialization statement related to this tag:

```
    init llvm.NoCFI                          {NoCFI}
```

Why is that being interpreted differently in two places?  What I'd expect to see is two separate tags (`osv.cfi.NoCFI` and `osv.threeClass.NoCFI`) that happen to both be applied to the same instructions.

**Comment by amstrnad** [comment]
Too many `NoCFI`s. The policy tags themselves are separate. That part is fine. 

The issue is that what instructions `init llvm.NoCFI` affects changes when policies are combined. Does that make sense?

**Comment by sbrookes** [comment]
We're pretty sure this is a `policy-tool` issue, not an `llvm` issue... correct?

**Comment by ccasin** [comment]
Sounds most likely to be a tagging tools issue.  It doesn't sound related to policy-tool.

**Comment by ccasin** [comment]
So narrowing this down:  The output of policy tool for `cfi-threeClass` looks right to me.  As far as I know, LLVM is not told to consider specific policies (it just outputs a bunch of info that the tagging tools may or may not use based on the poliy), so the problem can't be there.

I suspect a bug somewhere in tagging tools.  We'd expect the same `llvm.*` require statement to be handled identically in two different policies.  So: the fact that `llvm.NoCFI` is handled differently for `cfi` and `threeClass` sounds like a bug, and the fact that you get their intersection when you have both sounds like a second bug.

Of course, we have not written down a clear specification for these initialization statements, so it's not surprising for this kind of bug to arise.  And it's more evidence that we should get around to thinking more clearly about init statements.

**Comment by sbrookes** [comment]
The only place I can think of that will be relevant in tagging tools, and it looks pretty clean to me: https://github.com/draperlaboratory/hope-policy-engine/blob/master/tagging_tools/LLVMMetadataTagger.py#L202

**Comment by amstrnad** [comment]
It seems to me that what `llvm.NoCFI` is varies depending on what other `llvm.*` are used, and perhaps it should not.

**Comment by ccasin** [comment]
Agreed - it should not.  But perhaps we'll find there's some perfectly good reason it was build this way in the first place.  I will not have time to dig into this this week, unfortunately.
